### PR TITLE
Improve ActionException's for the Rest Dispatcher

### PR DESCRIPTION
**Pull request by BenDol**
This gives better understanding of issues during transport.
#### Exception Hierarchy:

|-- Exception
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`-- ActionException
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`-- RestActionException
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-- ActionResponseException
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`-- ActionDeserializationException
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`-- ActionSerializationException
##### [ActionException.java](https://github.com/BenDol/GWTP/blob/f61c072a9da756f1b82b0971bd9b316338368a55/gwtp-core/gwtp-dispatch-common-shared/src/main/java/com/gwtplatform/dispatch/shared/ActionException.java)

This exception now takes a `TypedAction<?>` (which has been made `Serializable`).

**Note:** There have been issues with variables in Exceptions for the RPC implementation, will need to know what the limitations are if there are any.
##### [RestActionException.java](https://github.com/BenDol/GWTP/blob/f61c072a9da756f1b82b0971bd9b316338368a55/gwtp-core/gwtp-dispatch-rest-shared/src/main/java/com/gwtplatform/dispatch/rest/shared/RestActionException.java)

Used as convenience for type casting the `TypedAction<?>` directly as `RestAction<?>`. Extends the `ActionException` class.
##### [ActionResponseException.java](https://github.com/BenDol/GWTP/blob/f61c072a9da756f1b82b0971bd9b316338368a55/gwtp-core/gwtp-dispatch-rest-shared/src/main/java/com/gwtplatform/dispatch/rest/shared/ActionResponseException.java)

Used in the response phase providing most of the `Response` information in the exception. `ActionDeserializationException` extends this class.
##### [ActionSerializationException.java](https://github.com/BenDol/GWTP/blob/f61c072a9da756f1b82b0971bd9b316338368a55/gwtp-core/gwtp-dispatch-rest-shared/src/main/java/com/gwtplatform/dispatch/rest/shared/ActionSerializationException.java)

Used in the serialization of an action request, mostly when a [SerializationException](https://github.com/ArcBees/GWTP/blob/master/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/serialization/SerializationException.java) is thrown.
##### [ActionDeserializationException.java](https://github.com/BenDol/GWTP/blob/f61c072a9da756f1b82b0971bd9b316338368a55/gwtp-core/gwtp-dispatch-rest-shared/src/main/java/com/gwtplatform/dispatch/rest/shared/ActionDeserializationException.java)

Used during the deserialization of a response from the server, only thrown when the response is unable to be deserialized by the object mapper. This class extends `ActionResponseException` providing the `Response` information with it.

### CODE REVIEW

**Comment by BenDol** [comment]
Added an example to the carstore sample application https://github.com/ArcBees/GWTP-Samples/pull/91

**Comment by Chris-V** [comment]
I'm not sure I can see the need to have both `ActionResponseException` and `ActionDeserializationException`.

Imo the action instance should be available through the deserialization as well.

File: gwtp-core/gwtp-dispatch-common-shared/src/main/java/com/gwtplatform/dispatch/shared/ActionException.java [hunk 1]
```diff
@@ -36,7 +36,10 @@ public ActionException(String message, Throwable cause) {
     }
 
     public ActionException(Throwable cause) {
-        super(cause.getMessage());
+        super(cause);
     }
 
+    public ActionException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace) {
```
Comment (Chris-V) [inline gwtp-core/gwtp-dispatch-common-shared/src/main/java/com/gwtplatform/dispatch/shared/ActionException.java]: I don't see the use of this overload

File: gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultBodyFactory.java [hunk 2]
```diff
@@ -127,11 +124,16 @@ private SerializedValue getSerializedValue(RestAction<?> action, Object object)
             Serialization serialization = findSerialization(bodyClass, contentTypes);
 
             if (serialization != null) {
-                return serialize(serialization, object, bodyClass, contentTypes);
+                try {
```
Comment (Chris-V) [inline gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultBodyFactory.java]: I think it was more readable with `try-catch` was part of `serialize`

File: gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultResponseDeserializer.java [hunk 3]
```diff
@@ -103,11 +101,16 @@ private boolean isSuccessStatusCode(Response response) {
             Serialization serialization = findSerialization(resultClass, contentType);
 
             if (serialization != null) {
-                return deserializeValue(serialization, resultClass, contentType, response.getText());
+                try {
```
Comment (Chris-V) [inline gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultResponseDeserializer.java]: Same here. It's more readable when part of `deserializeValue`

Comment (BenDol) [inline gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultResponseDeserializer.java]: I agree it was moved out when I was using the action as a parameter argument, but I removed it so I could move it back.

File: gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultBodyFactory.java (hunk 2 above)
Comment (BenDol) [inline gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultBodyFactory.java]: I agree it was moved out when I was using the action as a parameter argument, but I removed it so I could move it back.

File: gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultResponseDeserializer.java (hunk 3 above)
Comment (BenDol) [inline gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/DefaultResponseDeserializer.java]: Actaully this one needs the response data so it has to go outside it or the response needs to be passed in.

**Comment by BenDol** [comment]
Well a response exception is when there is a response problem not a deserialization problem, so if the server returns a unsuccessful status code it throws the ActionResponseException with the response information. The deserialization exception extends the response exception because its part of the response phase and has response information.

**Comment by BenDol** [comment]
This allows me to catch response specific errors and handle them in my exception handler. I.e. security exceptions where I notify the user of insufficient rights.

**Comment by BenDol** [comment]
Is this PR going to be used or reviewed to be used? Would be helpful for me.

**Comment by christiangoudreau** [comment]
Going to be reviewed, sorry about this!

**Comment by Chris-V** [comment]
It would be interesting to have `RestActionException` take the `RestAction<?>` in argument, and maybe the callback. I can see potential uses for automated retry mechanisms.

Also, in `DefaultResponseDeserializer`, I believe throwing a `BadStatusCodeException` (that extends `ActionResponseException`) would help prevent bad handling. ie: if someone tests for `ActionResponseException` before testing for `ActionDeserializationException`, then the first test will match and may prevent the second one for reaching.

`RestActionException` and `ActionResponseException` can be made abstract to help understanding they are top-level exceptions.

Can we get rid of the "Action" prefix. This may cause clashes with other exception names, though those prefixes annoy me to no end and I'm trying to get rid of them as time goes on ;)

**Comment by olafleur** [comment]
@BenDol can we help you with that PR ?

**Comment by BenDol** [comment]
Hey Chris, since this PR is not going to be merged can you help me understand how the new exception handling should work? I currently use this for catching my exceptions in my latest project and would need to rewrite it to work the way you have redesigned it. Thanks!

**Comment by dolnz** [comment]
Any update?

**Pull request closed without merging**